frontend/src/pages/releases/link-preview.mdx

```diff
@@ -0,0 +1,246 @@
+
+<img
+  src="/images/release_notes/link-preview/coursetable-arch-diagram.png"
+  alt="Diagram showing CourseTable's architecture. The user's browser sends a GET request to Vercel, which returns HTML/JS/CSS. The JS then makes GET requests to the server, which returns JSON. JSON and HTML combined gets rendered in the browser and turns into the final UI."
+  width={1000}
+/>
+
+Once the JavaScript loads, the user's browser starts making further requests to API endpoints at `api.coursetable.com`. This is our backend server that handles things like adding friends, storing worksheets, sending course data, etc. Unlike Vercel, it can react to whatever request in whichever way we want.
+
+So now we are in a deadlock situation.
+
+- Crawlers want nicely rendered HTML containing all data it needs without running any JavaScript.
+- Vercel only serves HTML that we uploaded; it can't generate HTML on the fly when a request comes in (that is called server-side rendering, or SSR—which our service doesn't provide).
+- We only ever upload one HTML file, because that's how CSR works. And even if we do generate one HTML for every single course, the problem remains that course-modal is a query string, and Vercel can't serve different HTML based on the query.
+- If Vercel doesn't want to dynamically generate HTML, we can modify the HTML after-the-fact using React in the user's browser, but crawlers don't run JavaScript.
+- We do have an API server that's fit for this kind of task, but bot traffic goes to `coursetable.com`, not to `api.coursetable.com`!
+
+So it seems no party is willing to collaborate and play nice. What gives?
+
+<img
+  src="/images/release_notes/link-preview/slack-generate-html.png"
+  alt="Slack conversation between Alex and Sida. Alex proposes to generate HTML the first time it is requested, but doubts if it would still be crazy."
+  width={600}
+/>
+
+Aside: You might ask, why don't we just change the URL to `https://coursetable.com/catalog/202403-10471`, so Vercel can serve the `catalog/202403-12309.html` file? Well there are still two problems: (1) there are probably hundreds of thousands of courses growing by several thousand every year; (2) course modals can be shown on other pages too, such as by clicking a course in your worksheet, which means… we have to have another set of HTML files that get served for `https://coursetable.com/worksheet/202403-10471`. That's a lot of HTML. We can't build so much HTML in reasonable time, and Vercel definitely has an upper bound to how much static assets we can store in it. This option was never considered or experimented with, but we did built infrastructure to statically pregenerate HTML for every URL at [#1536](https://github.com/coursetable/coursetable/pull/1536) (which Sida requested Alex's review on last year and he still hasn't…), in case it's useful for some future use case.
```

Comment:
```suggestion
Aside: You might ask, why don't we just change the URL to `https://coursetable.com/catalog/202403-10471`, so Vercel can serve the `catalog/202403-12309.html` file? Well there are still two problems: (1) there are hundreds of thousands of courses growing by several thousand every year; (2) course modals can be shown on other pages too, such as by clicking a course in your worksheet, which means… we have to have another set of HTML files that get served for `https://coursetable.com/worksheet/202403-10471`. That's a lot of HTML. We can't build so much HTML in reasonable time, and Vercel definitely has an upper bound to how much static assets we can store in it. This option was never considered or experimented with, but we did built infrastructure to statically pregenerate HTML for every URL at [#1536](https://github.com/coursetable/coursetable/pull/1536) (which Sida requested Alex's review on last year and he still hasn't…), in case it's useful for some future use case.
```

we r def over 100k now i think (should update landing page too) but can do a quick query to check

Reply:
Yeah looks like we are at 150,000